atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/TopListFilteredCounter.java

```diff
@@ -0,0 +1,123 @@
+    private static final SafeLogger log = SafeLoggerFactory.get(TopListFilteredCounter.class);
+
+    private final int maxSize;
+    private final TaggedMetricRegistry registry;
+    private final Function<T, MetricName> tagToMetricName;
+
+    private final Map<T, Counter> counters = new ConcurrentHashMap<>();
+    private final Set<T> reportedTags = ConcurrentHashMap.newKeySet();
+
+    private TopListFilteredCounter(
+            int maxSize, TaggedMetricRegistry registry, Function<T, MetricName> tagToMetricName) {
+        this.maxSize = maxSize;
+        this.registry = registry;
+        this.tagToMetricName = tagToMetricName;
+    }
+
+    static <T> TopListFilteredCounter<T> create(
+            int maxSize,
+            Duration initialDelay,
+            Duration resetInterval,
+            Function<T, MetricName> tagToMetricName,
+            TaggedMetricRegistry registry,
+            ScheduledExecutorService executor) {
+        TopListFilteredCounter<T> counter = new TopListFilteredCounter<>(maxSize, registry, tagToMetricName);
+        executor.scheduleAtFixedRate(
+                counter::resilientUpdateMetricsReporting,
+                initialDelay.toNanos(),
+                resetInterval.toNanos(),
+                TimeUnit.NANOSECONDS);
+        return counter;
+    }
+
+    void inc(T tag, long count) {
+        Counter counter = counters.computeIfAbsent(tag, unused -> new Counter());
+        counter.inc(count);
+    }
+
+    private void resilientUpdateMetricsReporting() {
+        try {
+            updateMetricsReporting();
+        } catch (RuntimeException e) {
+            log.error("Failed to update top listed counter metric", e);
+        }
+    }
+
+    private void updateMetricsReporting() {
+        Set<T> nextBatchOfTagsToReport = computeTopListTags();
+
+        deregisterTagsForMetricsProduction(Sets.difference(reportedTags, nextBatchOfTagsToReport));
+        registerTagsForMetricsProduction(Sets.difference(nextBatchOfTagsToReport, reportedTags));
+
+        reportedTags.clear();
+        reportedTags.addAll(nextBatchOfTagsToReport);
+    }
+
+    private void registerTagsForMetricsProduction(Set<T> tags) {
+        tags.forEach(this::registerTag);
+    }
+
+    private void deregisterTagsForMetricsProduction(Set<T> tags) {
+        tags.forEach(this::deregisterTag);
+    }
+
+    private void registerTag(T tag) {
+        MetricName metricName = tagToMetricName.apply(tag);
+        Counter counter = counters.get(tag);
+        // registers the counter only if no other counter is currently tracked for this metric name
+        registry.counter(metricName, () -> counter);
+    }
+
+    private void deregisterTag(T tag) {
+        MetricName metricName = tagToMetricName.apply(tag);
+        registry.remove(metricName);
+    }
+
+    private Set<T> computeTopListTags() {
+        Map<T, Long> counterSnapshots =
+                EntryStream.of(counters).mapValues(Counter::getCount).toMap();
```

Comment:
Note: should we filter out any that are at 0 or some low threshold? Otherwise we'll end up reporting pretty much everything at times on a bounce (across multiple top list refreshes)
And given that we have no consistent starting order, each refresh _could_ give a different result while all 0.

Reply:
Discussed offline: using a stable sort over the "tag -> counter#getCount" mitigates the issue. Adding a test for this is kind of awkward, so skipping that